ethcore/src/miner/filter_options.rs

```diff
@@ -0,0 +1,823 @@
+        let val = self.next_value()?;
+        match val {
+            Any | Eq(_) | ContractCreation => Ok(val),
+            _ => {
+                Err(M::Error::custom(
+                    "the sender filter only supports the `eq` and `action` operators",
+                ))
+            }
+        }
+    }
+    fn validate_value(&mut self) -> Result<FilterOperator<T>, M::Error> {
+        use self::FilterOperator::*;
+        let val = self.next_value()?;
+        match val {
+            Any | Eq(_) | GreaterThan(_) | LessThan(_) => Ok(val),
+            ContractCreation => {
+                Err(M::Error::custom(
+                    "the operator `action` is only supported by the receiver filter",
+                ))
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for FilterOptions {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct FilterOptionsVisitor;
+        impl<'de> Visitor<'de> for FilterOptionsVisitor {
+            type Value = FilterOptions;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                // "This Visitor expects to receive ..."
+                formatter.write_str("a map with one valid filter such as `sender`, `receiver`, `gas`, `gas_price`, `value` or `nonce`")
+            }
+
+            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'de>,
+            {
+                let mut filter = FilterOptions::default();
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "sender" => {
+                            filter.sender = map.validate_sender()?;
+                        },
+                        "receiver" => {
+                            filter.receiver = map.validate_receiver()?;
+                        },
+                        "gas" => {
+                            filter.gas = map.validate_value()?;
+                        },
+                        "gas_price" => {
+                            filter.gas_price = map.validate_value()?;
+                        },
+                        "value" => {
+                            filter.value = map.validate_value()?;
+                        },
+                        "nonce" => {
+                            filter.nonce = map.validate_value()?;
+                        },
+                        unknown => {
+                            return Err(M::Error::unknown_field(
+                                unknown,
+                                &["sender", "receiver", "gas", "gas_price", "value", "nonce"],
+                            ))
+                        }
+                    }
+                }
+
+                Ok(filter)
+            }
+        }
+
+        impl<'de, T: Deserialize<'de>> Deserialize<'de> for FilterOperator<T> {
```

Comment:
Do we really need custom `Deserialize` for `FilterOperator`? It seems to be tagged enum representation with a bunch of renames, no?